File: typescript/src/prng/permute.ts
/**
 * Feistel index permutation — a "lottery with no tumbler".
 *
 * A keyed pseudo-random **bijection** on `[0, n)`: `permute(i, n, key)` maps
 * each row index to a unique, scrambled "slot" without ever building an
 * array. This is what lets the streaming engine place a sorted quota plan
 * (`[A×70%, B×30%]`) into random-looking positions and get EXACT percentages
 * — because a bijection sends exactly `0.7n` indices into the first `0.7n`
 * slots — with O(1) memory and O(1) per row.
 *
 * Construction: a balanced Feistel network on the smallest `2^(2·half) ≥ n`,
 * with **cycle-walking** to fold the domain back down to exactly `[0, n)` for
 * non-power-of-two `n` (re-apply until the result is in range; expected < 4
 * iterations since the padded domain is < 4n). `unpermute` is the inverse
 * (needed by uniq's mixed-radix decode later).
 *
 * Determinism is cross-language: the round function uses only `Math.imul` and
 * 32-bit ops, and the key comes from cyrb128. Combining halves uses plain
 * arithmetic (not bit-shifts) so it stays exact for `n` up to ~2^52.
 */

import { cyrb128 } from './prng.js';

const ROUNDS = 4;

/** Derive a 32-bit round key for a permutation stream. */
export function permuteKey(seed: string, streamId: string): number {
  return cyrb128(`${seed}|perm|${streamId}`)[0];
}

/** Padded domain for `n`: two equal halves, `2^(2·half) ≥ n`. */
function halfSizeFor(n: number): number {
  const bits = Math.max(2, Math.ceil(Math.log2(n)));
  const half = Math.ceil(bits / 2);
  return 2 ** half;
}

/** 32-bit avalanche of the low half `r`, mixed with the round and key. */
function roundFn(r: number, round: number, key: number): number {
  let h = (r ^ Math.imul(round + 1, 0x9e3779b1)) >>> 0;
  h = Math.imul(h ^ (h >>> 16), 0x85ebca6b) >>> 0;
  h = Math.imul(h ^ (h >>> 13), 0xc2b2ae35) >>> 0;
  h = Math.imul((h ^ key) >>> 0, 0x27d4eb2f) >>> 0;
  return (h ^ (h >>> 16)) >>> 0;
}

function feistelForward(x: number, halfSize: number, key: number): number {
  let left = Math.floor(x / halfSize);
  let right = x % halfSize;
  for (let round = 0; round < ROUNDS; round++) {
    const mixed = roundFn(right, round, key) % halfSize;
    const nextRight = (left ^ mixed) >>> 0;
    left = right;
    right = nextRight;
  }
  return left * halfSize + right;
}

function feistelInverse(y: number, halfSize: number, key: number): number {
  let left = Math.floor(y / halfSize);
  let right = y % halfSize;
  for (let round = ROUNDS - 1; round >= 0; round--) {
    const prevRight = left;
    const mixed = roundFn(prevRight, round, key) % halfSize;
    const prevLeft = (right ^ mixed) >>> 0;
    left = prevLeft;
    right = prevRight;
  }
  return left * halfSize + right;
}

/** Bijection `[0, n) → [0, n)`: the scrambled slot for row `index`. */
export function permute(index: number, n: number, key: number): number {
  if (n <= 1) return 0;
  const halfSize = halfSizeFor(n);
  let x = index;
  do {
    x = feistelForward(x, halfSize, key);
  } while (x >= n);
  return x;
}

/** Inverse of `permute`: the row index that maps to `slot`. */
export function unpermute(slot: number, n: number, key: number): number {
  if (n <= 1) return 0;
  const halfSize = halfSizeFor(n);
  let x = slot;
  do {
    x = feistelInverse(x, halfSize, key);
  } while (x >= n);
  return x;
}

File: rust/src/prng/permute.rs
//! A shuffle you can evaluate at one position without performing it.
//!
//! This is what lets an exact quota be resolved row by row. Laying out a
//! `percent="20,80"` split is easy — twenty per cent of the slots, then eighty —
//! but the result would come out sorted, every `A` before every `B`. Shuffling
//! fixes that and normally requires the whole column in memory.
//!
//! A format-preserving permutation removes the requirement: a small Feistel
//! network over the index space is a bijection, so row `i` can ask which slot it
//! owns and get an answer consistent with every other row's, without any of them
//! existing.
//!
//! The cycle-walking loop is what keeps it exact for a size that is not a power
//! of two: the network works over a padded domain, and any result past the end
//! is fed back through until it lands inside. It terminates because the network
//! is a bijection on the padded space.

use super::{cyrb128, ushr};

const ROUNDS: i32 = 4;

/// A key private to one stream, so two columns shuffle independently.
pub fn key(seed: &str, stream_id: &str) -> i32 {
    cyrb128(&format!("{seed}|perm|{stream_id}"))[0]
}

/// The slot row `index` owns, among `n`.
pub fn apply(index: i32, n: i32, key: i32) -> i32 {
    if n <= 1 {
        return 0;
    }
    let half = half_size_for(n);
    let mut x = index;
    loop {
        x = forward(x, half, key);
        if x < n {
            return x;
        }
    }
}

/// The inverse: which row owns `slot`.
pub fn unapply(slot: i32, n: i32, key: i32) -> i32 {
    if n <= 1 {
        return 0;
    }
    let half = half_size_for(n);
    let mut x = slot;
    loop {
        x = inverse(x, half, key);
        if x < n {
            return x;
        }
    }
}

/// The padded domain: two equal halves whose product covers `n`.
fn half_size_for(n: i32) -> i32 {
    let bits = ((f64::from(n).ln() / std::f64::consts::LN_2).ceil() as i32).max(2);
    let half = (f64::from(bits) / 2.0).ceil() as i32;
    1 << half
}

/// The round function.
///
/// The mixing constants deliberately overflow and the shifts must be logical,
/// as Java's `>>>` is. An arithmetic shift here gives a different permutation
/// and the same seed lands on different rows.
fn round_fn(r: i32, round: i32, key: i32) -> i32 {
    // Each constant is written as the signed 32-bit pattern it is. `0x9e3779b1`
    // is above i32::MAX, so it is spelled as a u32 and cast — writing it bare
    // would not compile, and widening it to i64 would change the arithmetic.
    let mut h = r ^ (round + 1).wrapping_mul(0x9e3779b1_u32 as i32);
    h = (h ^ ushr(h, 16)).wrapping_mul(0x85ebca6b_u32 as i32);
    h = (h ^ ushr(h, 13)).wrapping_mul(0xc2b2ae35_u32 as i32);
    h = (h ^ key).wrapping_mul(0x27d4eb2f);
    h ^ ushr(h, 16)
}

fn forward(x: i32, half_size: i32, key: i32) -> i32 {
    let mut left = x / half_size;
    let mut right = x % half_size;
    for round in 0..ROUNDS {
        let mixed = ((round_fn(right, round, key) as u32) % (half_size as u32)) as i32;
        let next_right = left ^ mixed;
        left = right;
        right = next_right;
    }
    left * half_size + right
}

fn inverse(y: i32, half_size: i32, key: i32) -> i32 {
    let mut left = y / half_size;
    let mut right = y % half_size;
    for round in (0..ROUNDS).rev() {
        let prev_right = left;
        let mixed = ((round_fn(prev_right, round, key) as u32) % (half_size as u32)) as i32;
        let prev_left = right ^ mixed;
        left = prev_left;
        right = prev_right;
    }
    left * half_size + right
}
